### ISSUE TITLE
Build errors

### ISSUE THREAD

**Issue by chiefymuc** [rfkd/aircontrol#20]
Hello,

I'm on Raspberry Pi 2B v1.1 with raspbian bullseye.

I then installed libconfig++-dev:

```
pi@aircontrol:~/aircontrol $ sudo apt-get install libconfig++-dev
Reading package lists... Done
Building dependency tree... Done
Reading state information... Done
libconfig++-dev is already the newest version (1.5-0.4).
0 upgraded, 0 newly installed, 0 to remove and 0 not upgraded.
```

and as wiringPi is deprecated I installed it from https://github.com/WiringPi/WiringPi

```
cd
git clone https://github.com/WiringPi/WiringPi
cd WiringPi/
./build
```

WiringPi is now in /home/pi/WiringPi and I see also the /usr/local/lib/libwiringPi.so file and the libwiringPi.so.2.70

Then followed the instructions on here to install aircontrol but get a undefined reference error in combination with wiringPi. I have googled for hours, but no luck resolving this.

```pi@aircontrol:~/aircontrol $ make
g++ -c -std=c++14 -Wall -Wno-unused-result -Iinclude -MMD -MP -MF build/aircontrol.d -o build/aircontrol.o source/aircontrol.cpp
g++ -lconfig++ -lwiringPi  build/aircontrol.o build/Configuration.o build/InstanceLock.o build/Replay.o build/ReplayParameters.o build/Scan.o build/ScanParameters.o build/Target.o build/TargetParameters.o build/Task.o -o bin/aircontrol
/usr/bin/ld: build/aircontrol.o: in function `main':
aircontrol.cpp:(.text+0x77c): undefined reference to `wiringPiSetupGpio'
/usr/bin/ld: build/aircontrol.o: in function `Configuration::Configuration()':
aircontrol.cpp:(.text._ZN13ConfigurationC2Ev[_ZN13ConfigurationC5Ev]+0x38): undefined reference to `libconfig::Config::Config()'
/usr/bin/ld: build/aircontrol.o: in function `Configuration::~Configuration()':
aircontrol.cpp:(.text._ZN13ConfigurationD2Ev[_ZN13ConfigurationD5Ev]+0x1c): undefined reference to `libconfig::Config::~Config()'
/usr/bin/ld: build/Configuration.o: in function `Configuration::load()':
Configuration.cpp:(.text+0x8c): undefined reference to `libconfig::Config::readFile(char const*)'
/usr/bin/ld: build/Configuration.o: in function `libconfig::Config::lookup(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) const':
Configuration.cpp:(.text._ZNK9libconfig6Config6lookupERKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEE[_ZNK9libconfig6Config6lookupERKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEE]+0x28): undefined reference to `libconfig::Config::lookup(char const*) const'
/usr/bin/ld: build/Configuration.o:(.ARM.extab+0x44): undefined reference to `typeinfo for libconfig::ParseException'
/usr/bin/ld: build/Configuration.o:(.ARM.extab+0x48): undefined reference to `typeinfo for libconfig::FileIOException'
/usr/bin/ld: build/Configuration.o:(.ARM.extab+0x68): undefined reference to `typeinfo for libconfig::SettingNotFoundException'
/usr/bin/ld: build/Replay.o: in function `Replay::airReplay() const':
Replay.cpp:(.text+0x2a0): undefined reference to `pinMode'
/usr/bin/ld: Replay.cpp:(.text+0x31c): undefined reference to `digitalWrite'
/usr/bin/ld: Replay.cpp:(.text+0x350): undefined reference to `pinMode'
/usr/bin/ld: build/ReplayParameters.o: in function `libconfig::Setting::lookupValue(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, int&) const':
ReplayParameters.cpp:(.text._ZNK9libconfig7Setting11lookupValueERKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEERi[_ZNK9libconfig7Setting11lookupValueERKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEERi]+0x30): undefined reference to `libconfig::Setting::lookupValue(char const*, int&) const'
/usr/bin/ld: build/ReplayParameters.o: in function `bool Configuration::getValue<int>(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, int&) const':
ReplayParameters.cpp:(.text._ZNK13Configuration8getValueIiEEbNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEES6_RT_[_ZNK13Configuration8getValueIiEEbNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEES6_RT_]+0x4c): undefined reference to `libconfig::Config::getRoot() const'
/usr/bin/ld: ReplayParameters.cpp:(.text._ZNK13Configuration8getValueIiEEbNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEES6_RT_[_ZNK13Configuration8getValueIiEEbNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEES6_RT_]+0x68): undefined reference to `libconfig::Setting::operator[](char const*) const'
/usr/bin/ld: build/ReplayParameters.o:(.ARM.extab.text._ZNK13Configuration8getValueIiEEbNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEES6_RT_[_ZNK13Configuration8getValueIiEEbNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEES6_RT_]+0x20): undefined reference to `typeinfo for libconfig::SettingNotFoundException'
/usr/bin/ld: build/Scan.o: in function `Scan::airScan()':
Scan.cpp:(.text+0x320): undefined reference to `pinMode'
/usr/bin/ld: Scan.cpp:(.text+0x378): undefined reference to `digitalRead'
/usr/bin/ld: build/Target.o: in function `Target::airControl() const':
Target.cpp:(.text+0x3c8): undefined reference to `pinMode'
/usr/bin/ld: Target.cpp:(.text+0x4b8): undefined reference to `digitalWrite'
/usr/bin/ld: Target.cpp:(.text+0x504): undefined reference to `pinMode'
/usr/bin/ld: build/Target.o: in function `Target::sendAirCommandManchester() const':
Target.cpp:(.text+0x670): undefined reference to `digitalWrite'
/usr/bin/ld: Target.cpp:(.text+0x6b0): undefined reference to `digitalWrite'
/usr/bin/ld: Target.cpp:(.text+0x6f0): undefined reference to `digitalWrite'
/usr/bin/ld: Target.cpp:(.text+0x738): undefined reference to `digitalWrite'
/usr/bin/ld: Target.cpp:(.text+0x784): undefined reference to `digitalWrite'
/usr/bin/ld: build/Target.o:Target.cpp:(.text+0x7cc): more undefined references to `digitalWrite' follow
/usr/bin/ld: build/TargetParameters.o: in function `libconfig::Setting::lookupValue(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >&) const':
TargetParameters.cpp:(.text._ZNK9libconfig7Setting11lookupValueERKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEERS6_[_ZNK9libconfig7Setting11lookupValueERKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEERS6_]+0x30): undefined reference to `libconfig::Setting::lookupValue(char const*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >&) const'
/usr/bin/ld: build/TargetParameters.o: in function `bool Configuration::getValue<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >&) const':
TargetParameters.cpp:(.text._ZNK13Configuration8getValueINSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEEEEbS6_S6_RT_[_ZNK13Configuration8getValueINSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEEEEbS6_S6_RT_]+0x4c): undefined reference to `libconfig::Config::getRoot() const'
/usr/bin/ld: TargetParameters.cpp:(.text._ZNK13Configuration8getValueINSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEEEEbS6_S6_RT_[_ZNK13Configuration8getValueINSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEEEEbS6_S6_RT_]+0x68): undefined reference to `libconfig::Setting::operator[](char const*) const'
/usr/bin/ld: build/TargetParameters.o:(.ARM.extab.text._ZNK13Configuration8getValueINSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEEEEbS6_S6_RT_[_ZNK13Configuration8getValueINSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEEEEbS6_S6_RT_]+0x20): undefined reference to `typeinfo for libconfig::SettingNotFoundException'
/usr/bin/ld: build/Task.o: in function `Task::isValidGpioPin(unsigned char)':
Task.cpp:(.text+0xb0): undefined reference to `piBoardRev'
collect2: error: ld returned 1 exit status
make: *** [Makefile:39: bin/aircontrol] Error 1
```

**Comment by rfkd** [comment]
Hey,

looks like neither `libconfig++` nor `wiringPi` can be found on your system. Meanwhile both libraries can be installed via APT, so run the following commands as root (or via `sudo`):
```
# apt update
# apt install libconfig++9v5 wiringPi
```

Afterwards try to build `aircontrol` again.

If it still doesn't work you may post the output of the following commands as root:
```
# dpkg -l | grep -P "wiringpi|libconfig\+\+"
# ldconfig
# ldconfig -p | grep -P "wiringPi|config\+\+"
# ldconfig -v 2>/dev/null | grep ^/
```

**Comment by chiefymuc** [comment]
Hi and thanks for your answer!

I went through the steps, unfortunately without success. Here is the output:

`
pi@aircontrol:~ $ sudo apt update
Get:1 http://archive.raspberrypi.org/debian bullseye InRelease [23.6 kB]
Get:2 http://raspbian.raspberrypi.org/raspbian bullseye InRelease [15.0 kB]
Get:3 http://raspbian.raspberrypi.org/raspbian bullseye/main armhf Packages [13.2 MB]
Get:4 http://archive.raspberrypi.org/debian bullseye/main armhf Packages [257 kB]
Fetched 13.5 MB in 15s (903 kB/s)
Reading package lists... Done
Building dependency tree... Done
Reading state information... Done
10 packages can be upgraded. Run 'apt list --upgradable' to see them.

pi@aircontrol:~ $ sudo apt install libconfig++9v5
Reading package lists... Done
Building dependency tree... Done
Reading state information... Done
libconfig++9v5 is already the newest version (1.5-0.4).
libconfig++9v5 set to manually installed.
0 upgraded, 0 newly installed, 0 to remove and 10 not upgraded.

pi@aircontrol:~ $ sudo apt install wiringPi
Reading package lists... Done
Building dependency tree... Done
Reading state information... Done
E: Unable to locate package wiringPi

pi@aircontrol:~ $ cd aircontrol/
pi@aircontrol:~/aircontrol $ sudo make
g++ -c -std=c++14 -Wall -Wno-unused-result -Iinclude -MMD -MP -MF build/aircontrol.d -o build/aircontrol.o source/aircontrol.cpp
g++ -lconfig++ -lwiringPi  -L/usr/local/lib build/aircontrol.o build/Configuration.o build/InstanceLock.o build/Replay.o build/ReplayParameters.o build/Scan.o build/ScanParameters.o build/Target.o build/TargetParameters.o build/Task.o -o bin/aircontrol
/usr/bin/ld: build/aircontrol.o: in function `main':
aircontrol.cpp:(.text+0x77c): undefined reference to `wiringPiSetupGpio'
/usr/bin/ld: build/aircontrol.o: in function `Configuration::Configuration()':
aircontrol.cpp:(.text._ZN13ConfigurationC2Ev[_ZN13ConfigurationC5Ev]+0x38): undefined reference to `libconfig::Config::Config()'
/usr/bin/ld: build/aircontrol.o: in function `Configuration::~Configuration()':
aircontrol.cpp:(.text._ZN13ConfigurationD2Ev[_ZN13ConfigurationD5Ev]+0x1c): undefined reference to `libconfig::Config::~Config()'
/usr/bin/ld: build/Configuration.o: in function `Configuration::load()':
Configuration.cpp:(.text+0x8c): undefined reference to `libconfig::Config::readFile(char const*)'
/usr/bin/ld: build/Configuration.o: in function `libconfig::Config::lookup(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) const':
Configuration.cpp:(.text._ZNK9libconfig6Config6lookupERKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEE[_ZNK9libconfig6Config6lookupERKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEE]+0x28): undefined reference to `libconfig::Config::lookup(char const*) const'
/usr/bin/ld: build/Configuration.o:(.ARM.extab+0x44): undefined reference to `typeinfo for libconfig::ParseException'
/usr/bin/ld: build/Configuration.o:(.ARM.extab+0x48): undefined reference to `typeinfo for libconfig::FileIOException'
/usr/bin/ld: build/Configuration.o:(.ARM.extab+0x68): undefined reference to `typeinfo for libconfig::SettingNotFoundException'
/usr/bin/ld: build/Replay.o: in function `Replay::airReplay() const':
Replay.cpp:(.text+0x2a0): undefined reference to `pinMode'
/usr/bin/ld: Replay.cpp:(.text+0x31c): undefined reference to `digitalWrite'
/usr/bin/ld: Replay.cpp:(.text+0x350): undefined reference to `pinMode'
/usr/bin/ld: build/ReplayParameters.o: in function `libconfig::Setting::lookupValue(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, int&) const':
ReplayParameters.cpp:(.text._ZNK9libconfig7Setting11lookupValueERKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEERi[_ZNK9libconfig7Setting11lookupValueERKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEERi]+0x30): undefined reference to `libconfig::Setting::lookupValue(char const*, int&) const'
/usr/bin/ld: build/ReplayParameters.o: in function `bool Configuration::getValue<int>(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, int&) const':
ReplayParameters.cpp:(.text._ZNK13Configuration8getValueIiEEbNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEES6_RT_[_ZNK13Configuration8getValueIiEEbNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEES6_RT_]+0x4c): undefined reference to `libconfig::Config::getRoot() const'
/usr/bin/ld: ReplayParameters.cpp:(.text._ZNK13Configuration8getValueIiEEbNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEES6_RT_[_ZNK13Configuration8getValueIiEEbNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEES6_RT_]+0x68): undefined reference to `libconfig::Setting::operator[](char const*) const'
/usr/bin/ld: build/ReplayParameters.o:(.ARM.extab.text._ZNK13Configuration8getValueIiEEbNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEES6_RT_[_ZNK13Configuration8getValueIiEEbNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEES6_RT_]+0x20): undefined reference to `typeinfo for libconfig::SettingNotFoundException'
/usr/bin/ld: build/Scan.o: in function `Scan::airScan()':
Scan.cpp:(.text+0x320): undefined reference to `pinMode'
/usr/bin/ld: Scan.cpp:(.text+0x378): undefined reference to `digitalRead'
/usr/bin/ld: build/Target.o: in function `Target::airControl() const':
Target.cpp:(.text+0x3c8): undefined reference to `pinMode'
/usr/bin/ld: Target.cpp:(.text+0x4b8): undefined reference to `digitalWrite'
/usr/bin/ld: Target.cpp:(.text+0x504): undefined reference to `pinMode'
/usr/bin/ld: build/Target.o: in function `Target::sendAirCommandManchester() const':
Target.cpp:(.text+0x670): undefined reference to `digitalWrite'
/usr/bin/ld: Target.cpp:(.text+0x6b0): undefined reference to `digitalWrite'
/usr/bin/ld: Target.cpp:(.text+0x6f0): undefined reference to `digitalWrite'
/usr/bin/ld: Target.cpp:(.text+0x738): undefined reference to `digitalWrite'
/usr/bin/ld: Target.cpp:(.text+0x784): undefined reference to `digitalWrite'
/usr/bin/ld: build/Target.o:Target.cpp:(.text+0x7cc): more undefined references to `digitalWrite' follow
/usr/bin/ld: build/TargetParameters.o: in function `libconfig::Setting::lookupValue(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >&) const':
TargetParameters.cpp:(.text._ZNK9libconfig7Setting11lookupValueERKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEERS6_[_ZNK9libconfig7Setting11lookupValueERKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEERS6_]+0x30): undefined reference to `libconfig::Setting::lookupValue(char const*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >&) const'
/usr/bin/ld: build/TargetParameters.o: in function `bool Configuration::getValue<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >&) const':
TargetParameters.cpp:(.text._ZNK13Configuration8getValueINSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEEEEbS6_S6_RT_[_ZNK13Configuration8getValueINSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEEEEbS6_S6_RT_]+0x4c): undefined reference to `libconfig::Config::getRoot() const'
/usr/bin/ld: TargetParameters.cpp:(.text._ZNK13Configuration8getValueINSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEEEEbS6_S6_RT_[_ZNK13Configuration8getValueINSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEEEEbS6_S6_RT_]+0x68): undefined reference to `libconfig::Setting::operator[](char const*) const'
/usr/bin/ld: build/TargetParameters.o:(.ARM.extab.text._ZNK13Configuration8getValueINSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEEEEbS6_S6_RT_[_ZNK13Configuration8getValueINSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEEEEbS6_S6_RT_]+0x20): undefined reference to `typeinfo for libconfig::SettingNotFoundException'
/usr/bin/ld: build/Task.o: in function `Task::isValidGpioPin(unsigned char)':
Task.cpp:(.text+0xb0): undefined reference to `piBoardRev'
collect2: error: ld returned 1 exit status
make: *** [Makefile:43: bin/aircontrol] Error 1

pi@aircontrol:~/aircontrol $ sudo dpkg -l | grep -P "wiringpi|libconfig\+\+"
ii  libconfig++-dev:armhf                1.5-0.4                          armhf        parsing/manipulation of structured config files (C++ development)
ii  libconfig++9v5:armhf                 1.5-0.4                          armhf        parsing/manipulation of structured configuration files (C++ binding)

pi@aircontrol:~/aircontrol $ sudo ldconfig

pi@aircontrol:~/aircontrol $ sudo ldconfig -p | grep -P "wiringPi|config\+\+"
        libwiringPiDev.so (libc6,hard-float) => /usr/local/lib/libwiringPiDev.so
        libwiringPiDev.so (libc6,hard-float) => /lib/libwiringPiDev.so
        libwiringPi.so (libc6,hard-float) => /usr/local/lib/libwiringPi.so
        libwiringPi.so (libc6,hard-float) => /lib/libwiringPi.so
        libconfig++.so.9 (libc6,hard-float) => /lib/arm-linux-gnueabihf/libconfig++.so.9
        libconfig++.so (libc6,hard-float) => /lib/arm-linux-gnueabihf/libconfig++.so

pi@aircontrol:~/aircontrol $ sudo ldconfig -v 2>/dev/null | grep ^/
/lib/arm-linux-gnueabihf:
/usr/lib/arm-linux-gnueabihf/libfakeroot:
/usr/local/lib:
/lib:

`

**Comment by rfkd** [comment]
Thanks for your feedback.

As your setup looks correct I've tried it out on a Bullseye installation myself and was able to reproduce your problem. There was a bug in the linker command line argument order which was accepted at least until GCC 8 (used in Raspbian Buster), but not anymore by GCC 10 (used in Raspbian Bullseye). The bug is fixed and a new release in version [0.2.1](https://github.com/rfkd/aircontrol/releases/tag/0.2.1) is created.

By the way, there is a small typo in your command when trying to install wiringPi - it needs to be `apt install wiringpi`, not `apt install wiringPi` (small `p`). So install wiringPi via APT, update your aircontrol clone to the latest `master` (or `develop`) and try to rebuild.

**Comment by chiefymuc** [comment]
This worked, thank you for the fix!

However `apt install wiringpi` still came back with an error.
